File: server/schemas/knowledge.py
"""Schemas for Knowledge management."""

from __future__ import annotations

from typing import Any
from datetime import datetime

from pydantic import BaseModel


class KnowledgeSourceCreate(BaseModel):
    name: str
    source_type: str  # document | faq | structured_table | kv_entity
    domain: str = "default"
    tenant_id: str = "default"
    metadata: dict[str, Any] | None = None


class KnowledgeSourceOut(BaseModel):
    id: str
    name: str
    source_type: str
    source_uri: str
    domain: str
    tenant_id: str
    status: str
    metadata_: Any
    chunk_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class KVEntityCreate(BaseModel):
    """For fast-answer channel: structured key-value entities."""
    source_id: str
    entity_key: str  # e.g. "市民服务中心电话"
    content: str     # e.g. "0571-12345678"
    domain: str = "default"
    metadata: dict[str, Any] | None = None


class FAQCreate(BaseModel):
    """For FAQ entries."""
    source_id: str
    question: str
    answer: str
    domain: str = "default"
    metadata: dict[str, Any] | None = None


class RetrievalRequest(BaseModel):
    query: str
    domain: str | None = None
    top_k: int = 5
    use_fast_channel: bool = True
    use_rag_channel: bool = True


class RetrievalHit(BaseModel):
    chunk_id: str
    source_id: str
    source_name: str
    content: str
    score: float
    page: int | None = None
    paragraph: int | None = None
    line_start: int | None = None
    line_end: int | None = None
    channel: str  # fast | rag


class RetrievalResponse(BaseModel):
    hits: list[RetrievalHit] = []
    fast_answer: str | None = None
    query: str = ""
    latency_ms: float = 0.0

File: console/src/types/knowledge.ts
export interface KnowledgeSource {
  id: string;
  name: string;
  source_type: 'document' | 'faq' | 'structured_table' | 'kv_entity';
  source_uri: string;
  domain: string;
  tenant_id: string;
  status: 'pending' | 'processing' | 'ready' | 'error';
  metadata: Record<string, unknown> | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeChunk {
  id: string;
  source_id: string;
  content: string;
  entity_key: string | null;
  domain: string;
  chunk_index: number;
  created_at: string;
}

export interface KnowledgeSourceCreate {
  name: string;
  source_type: string;
  domain?: string;
  tenant_id?: string;
}

export interface RetrievalRequest {
  query: string;
  domain?: string;
  top_k?: number;
  use_fast_channel?: boolean;
  use_rag_channel?: boolean;
}

export interface RetrievalHit {
  chunk_id: string;
  content: string;
  score: number;
  source: string;
  channel: string;
}

export interface RetrievalResponse {
  hits: RetrievalHit[];
  fast_answer: string | null;
  query: string;
}
